File: Algorithms/0752_Open_the_Lock/Java/OpentheLockSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> cache = new HashSet<>();
        for (String s : deadends) cache.add(s);
        if (cache.contains(target) || cache.contains("0000")) return -1;
        int step = 0;
        Queue<String> queue = new ArrayDeque<>();
        queue.add("0000");

        while (!queue.isEmpty()) {
            int count = queue.size();
            while (count-- > 0) {
                String cur = queue.poll();
                if (cur.equals(target)) return step;
                for (int i = 0; i < cur.length(); i++) {
                    int tempNum = Character.getNumericValue((int) cur.charAt(i));
                    StringBuilder t1, t2;
                    t1 = new StringBuilder(cur.substring(0, i)).append(String.valueOf((10 + tempNum + 1) % 10)).append(cur.substring(i + 1));
                    t2 = new StringBuilder(cur.substring(0, i)).append(String.valueOf((10 + tempNum - 1) % 10)).append(cur.substring(i + 1));
                    if (!cache.contains(t1.toString())) {
                        queue.add(t1.toString());
                        cache.add(t1.toString());
                    }
                    if (!cache.contains(t2.toString())) {
                        queue.add(t2.toString());
                        cache.add(t2.toString());
                    }
                }
            }
            step++;
        }
        return -1;
    }
}

File: Algorithms/0752_Open_the_Lock/Python/Open_the_Lock_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def openLock(self, deadends, target):
        if target in deadends or '0000' in deadends: return -1
        deadends = set(deadends)

        queue = [('0000', 0)]
        while queue:
            cur_number, cur_step = queue.pop(0)
            if cur_number == target: return cur_step
            for i in range(4):
                temp_number = int(cur_number[i])
                temp1 = cur_number[:i] + str((temp_number + 1) % 10) + cur_number[i + 1:]
                temp2 = cur_number[:i] + str((temp_number - 1) % 10) + cur_number[i + 1:]
                if temp1 not in deadends:
                    queue.append((temp1, cur_step + 1))
                    deadends.add(temp1)
                if temp2 not in deadends:
                    queue.append((temp2, cur_step + 1))
                    deadends.add(temp2)
        return -1






